feat: tweak weekly tests

DavePearce opened the pull request:
Ready to merge.

DavePearce committed:
sepearate our precompile call tests

Since the precompile call tests have a different dynamic from other
weekly tests, have moved all of them into the group that was previously
the weekly ecpairing suite.  That way, the parallelism for this group
can be dramatically increased without causing other (costly and
unrelated) weekly tests from failing.

DavePearce committed:
add pull_request trigger

The purpose of this is just for testing to let me run the workflow prior
to it being merged in (since it hasn't been run before).

DavePearce committed:
Tweaking JVM memory config

The challenge with using a high level of parallelism is that we end up
creating a large number of JVMs.  With a minimum heap size of 4GB, we
completely max out memory at 32 processes.  That's ignoring memory
required for the OS and/or go-corset itself.

However, most of these short-lived tests do not need anything like 4GB
to complete.  Setting a minimum heap size of 512MB means the average
test consumes much less RAM, whilst keeping the maximum heap size at 8GB
means there is room for them to expand as needed.

The final challenge with increasing the degree of parallelism is
go-corset itself.  This also uses a large number of cores, even for a
single test.  Therefore, there is a careful balance needed to make
effective use of all available cores.

DavePearce committed:
Enable random sampling by day-of-month

This enables random sampling using the day of the month as a seed for
the RNG.  This means that, in principle, we can recreate any failures by
looking up on which day the test failed.  At the same time, we benefit
from wider testing over a period of time.

A default sample size is given individually for each test, and this can
be overridden using a single environment variable
PRC_CALLTESTS_SAMPLE_SIZE.  Hopefully, this provides the level of
flexibility needed.

amkCha commented on .github/workflows/gradle-weekly-tests.yml:
```diff
@@ -11,9 +11,8 @@ concurrency:
 
 jobs:
-  weekly-tests:
+  weekly-replay-tests:
```
Are they replay ? Just checking

DavePearce replied:
Hmmm, I thought they were.  Based on what Olivier said last night though, you have a point.  I will check that now.

DavePearce replied:
Done.  No longer marked as replay.

amkCha commented on .github/workflows/gradle-prc-call-tests.yml:
```diff
@@ -0,0 +1,48 @@
+name: Weekly Precompile Call Tests
```
It was probably done on purpose, but do you think we should separate those weeklys from the others ? It's having us manage another flow and another file. They could run on the same github-action with 2 runners in // though and upload 2 different reports

DavePearce replied:
So, the reason I did this is that I believe (perhaps incorrectly) that the test profile for the weeklys is quite different from the PRC call tests.  However, I will look at that now, since I previously thought the other weeklys were all replays.  Key point being that we cannot run the replays with `X_PARALLELISM=26`.

DavePearce replied:
Ok, so I think you are right here --- nice one.  We can potentially do the whole lot under a single "weekly" tag.  I will try to set that up now.

DavePearce replied:
Done. Lets see if it works.

DavePearce committed:
consolidate prcCallTests => weeklyTests

This consolidates all of the prcCallTests into the existing weeklyTests
group.  This makes sense as they all have roughly similar test profiles.
The weeklyTests now has a much more aggressive degree of parallelism.

In addition, this sets more suitable default sample sizes for the PRC
calltests.  Specifically, defaults are set to approx 10% of the overall
parameter space.

DavePearce committed:
split weekly tests into two jobs

This splits the weekly tests up into two distinct jobs so they can run
in parallel.  This hopefully should allow for a larger sampling size.
